core/src/main/java/org/apache/calcite/rel/externalize/RelJson.java

```diff
@@ -391,13 +385,17 @@ RexNode toRex(RelInput relInput, Object o) {
       }
       if (map.containsKey("literal")) {
         final Object literal = map.get("literal");
-        final SqlTypeName sqlTypeName =
-            Util.enumVal(SqlTypeName.class, (String) map.get("type"));
+        final RelDataType type = toType(typeFactory, map.get("type"));
```

Comment:
this seems to be changing what the "type" was supposed to contain. it looks like earlier, it was something like
`BOOLEAN` ; but now it's a more complete `toJson(node.getType())` so parsing of a record written by an earlier writer will result in a ClassCastException at this line because toType is guessing that it's a Map

IMHO before we should be getting into things which involve version compatibility - the json should get some version field or something; to avoid writing a convoluted reader

I don't know how many people are there who are actually using this - but how about simply dropping backward compat for now?
Then all of those "o instanceof Boolean" and it's friends could go away as well.

Reply:
@kgyrtkirk , thanks for catching that. I have uploaded a follow-up fix and a test to verify that it all works as expected.
You will see that the fix is quite simple. I am not sure how widely used the JSON reader/writer are, but I believe we should not break backwards compatibility since it is quite straightforward to keep it. Besides, versioning sounds like a good idea moving forward, but maybe not needed for this change? Could you take a look again?

Reply:
looks good
:+1: